Announce error-stack v0.2 (#1146)
hashintel/hash: sites/hashdev/src/_pages/blog/10_error-stack-update-0-2.mdx

---
authorPhoto: blog/authors/tim-diekmann.jpg
postPhoto: blog/10_error-stack-update-0-2/Cover.jpg
title: "Announcing error-stack v0.2"
subtitle: "Redesigned and customizable outputs, related errors and multiple sources, and much more"
author: "Tim Diekmann"
jobTitle: "Platform Engineer, HASH"
date: "2022-10-03"
---

> Back in June [we announced the first version](https://hash.dev/blog/announcing-error-stack) of `error-stack`, our context-aware error library. We developed this to help address a lot of challenges we encountered when implementing error handling in our simulation engine [LINK]. In the time since then we’ve started using it in our entity-graph query layer [LINK?] that we’re building as a datastore for HASH [LINK?]. Across that development we’ve learned a lot of lessons, and in joint efforts with open-source contributors we’re ready to release a new version of the library.

Author: Do we want to link the engine to `hash.ai` or to GH?

Reviewer: I would link it to https://hash.ai/platform/engine which links to GH anyway – @nonparibus to confirm

Author: Added link in 74a3f56b248df2ef0fd7bc2e4adf08004bdddf65

> Back in June [we announced the first version](https://hash.dev/blog/announcing-error-stack) of `error-stack`, our context-aware error library. We developed this to help address a lot of challenges we encountered when implementing error handling in our simulation engine [LINK]. In the time since then we’ve started using it in our entity-graph query layer [LINK?] that we’re building as a datastore for HASH [LINK?]. Across that development we’ve learned a lot of lessons, and in joint efforts with open-source contributors we’re ready to release a new version of the library.

Author: Do we want to link the entity-graph query layer?

Reviewer: What would we link to here?

Author: I think mainly to the dev/graph work, but as this is a rather temporary link target, we should probably just don't add a link, as it's basically included in `HASH` a few words later.

Reviewer 2: True, without a better place to link to I definitely wouldn't do it now.

> Back in June [we announced the first version](https://hash.dev/blog/announcing-error-stack) of `error-stack`, our context-aware error library. We developed this to help address a lot of challenges we encountered when implementing error handling in our simulation engine [LINK]. In the time since then we’ve started using it in our entity-graph query layer [LINK?] that we’re building as a datastore for HASH [LINK?]. Across that development we’ve learned a lot of lessons, and in joint efforts with open-source contributors we’re ready to release a new version of the library.

Author: Do we want to link the Graph API?

Reviewer: I would just link "HASH" to https://hash.ai – @nonparibus to confirm

Author: Added link in 74a3f56b248df2ef0fd7bc2e4adf08004bdddf65

> Back in June [we announced the first version](https://hash.dev/blog/announcing-error-stack) of `error-stack`, our context-aware error library. We developed this to help address a lot of challenges we encountered when implementing error handling in our simulation engine [LINK]. In the time since then we’ve started using it in our entity-graph query layer [LINK?] that we’re building as a datastore for HASH [LINK?]. Across that development we’ve learned a lot of lessons, and in joint efforts with open-source contributors we’re ready to release a new version of the library.

Reviewer: Optional trimming of words
```suggestion
Back in June [we announced the first version](https://hash.dev/blog/announcing-error-stack) of `error-stack`, our context-aware error library. We developed this to help address challenges we encountered when implementing error handling in our simulation engine [LINK]. Since then we’ve started using it in our entity-graph query layer [LINK?] that we’re building as a datastore for HASH [LINK?]. Across that development we’ve learned a lot of lessons, and in joint efforts with open-source contributors we’re ready to release a new version of the library.
```

Reviewer 3: Re-captured as a suggestion as part of https://github.com/hashintel/hash/pull/1146/files#r985691554 so marking this suggest as resolved.

> Almost every week since our last release, we’ve been fixing user-reported bugs, reviewing open-source contributor’s pull requests, and adding new features. [Version 0.2](https://github.com/hashintel/hash/tree/error-stack%400.2.0/packages/libs/error-stack) is available now and includes important changes from `std`, a completely redesigned output with new customization options, support for related errors and multiple error sources, and a number of other improvements.

Reviewer: Optional. I feel like "almost" something lacks impact. If we want to imply continual progression we could say "Since our last release we've been [steadily|continually|some other word] fixing ..." .

```suggestion
Since our last release we’ve been fixing user-reported bugs, reviewing open-source contributor’s pull requests, and adding new features. [Version 0.2](https://github.com/hashintel/hash/tree/error-stack%400.2.0/packages/libs/error-stack) is available now and includes important changes from `std`, a completely redesigned output with new customization options, support for related errors and multiple error sources, and a number of other improvements.
```

## Changes from `std` in v0.2

- The `Provider` API has landed and replaces our vendored-in implementation.

> - `Backtrace` was stabilized with Rust 1.65(🎉), we now support `Backtrace`s on a non-nightly channel starting with 1.65-beta.

Reviewer: ```suggestion
- `Backtrace` was stabilized with Rust 1.65(🎉): we now support `Backtrace`s on a non-nightly channel starting with 1.65-beta.
```

> - If the root context already contains a backtrace, no new `Backtrace` is captured. Otherwise, backtraces are captured automatically when `RUST_BACKTRACE`.
>   or `RUST_LIB_BACKTRACE` env-vars are set.

Reviewer: ```suggestion
- If the root context already contains a backtrace, no new `Backtrace` is captured. Otherwise, backtraces are captured automatically when `RUST_BACKTRACE` or `RUST_LIB_BACKTRACE` env-vars are set.
```

```rust
// Note: neither `Debug`, nor `Display` is implemented
 struct Suggestion(&'static str);

 Report::install_debug_hook::<Suggestion>(|value, context| {
     context.push_body(format!("suggestion: {}", value.0));
 });

 let report = read("config.txt").attach(Suggestion("check if `config.txt` exists"));
```

```text
no such file or directory
 ├╴read.rs:10:5
 ╰╴suggestion: check if `config.txt` exists
```

Report outputs now consist of two sections: a `Body` and an `Appendix`. The `Body` provides top-level information while the `Appendix` includes additional, more verbose details, like a backtrace. Hooks can be used to write to both of these sections.

```text
experiment error: could not run experiment
 ├╴examples/demo.rs:67:17
 ├╴backtrace with 19 frames (3)
 ├╴experiment 3 has no valid description
 ╰╴unable to set up experiments

 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

>  Backtrace No. 1

Author: This is a direct copy/paste from the output

>  Backtrace No. 1

Reviewer 2: But are there more backtraces below which were cut out? 

I agree we want representative examples, but I think we should make effort to pick cases that actually demonstrate the thing we're on about.

If this _was_ the full output then I'm actually confused by it and I think we have a bug..

>  Backtrace No. 1

Author: Oh, that was trimmed, I haven't noticed that. Sorry, pushed a fix for that 9174b86038a433a193ee202bef1ecd7f8ff4e2d2

>  Backtrace No. 1

Author: The output is the same as below, but with backtraces enabled

    0: std::backtrace_rs::backtrace::libunwind::trace
              at /rustc/8b705839cd656d202e920efa8769cbe43a5ee269/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
    1: std::backtrace_rs::backtrace::trace_unsynchronized
              at /rustc/8b705839cd656d202e920efa8769cbe43a5ee269/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
    2: std::backtrace::Backtrace::create
              at /rustc/8b705839cd656d202e920efa8769cbe43a5ee269/library/std/src/backtrace.rs:333:13
    3: core::ops::function::FnOnce::call_once
              at /rustc/8b705839cd656d202e920efa8769cbe43a5ee269/library/core/src/ops/function.rs:251:5
    ...
    18: _main
```

For a full guide to using hooks, please refer to [the `fmt` module documentation](https://docs.rs/error-stack/0.2.0/error_stack/fmt/index.html).

## Related errors and multiple error sources

> `error-stack` v0.2 introduces support for related errors and multiple error sources, which can be invaluable when multiple errors occur at the same time, e.g. in loops or in multi-threaded applications. `Report`s are no longer a linear stream of events. You can now create a tree of different `Report`s, enabling the capture of multiple simultaneous errors without loosing information.

Reviewer: ```suggestion
`error-stack` v0.2 introduces support for related errors and multiple error sources, which can be invaluable when multiple errors occur at the same time, e.g. in loops or in multi-threaded applications. `Report`s are no longer a linear stream of events. You can now create a tree of different `Report`s, enabling the capture of multiple simultaneous errors without losing information.
```

This is also shown in the output:

> ```text

Author: I did some testing in the draft at Notion, but wasn't satisfied with a solution.

> ```text

Reviewer 2: I think it's fine as is, I posted this before looking at the rendered view and I think they're fine without colour :)

Error: Experiment error: Could not run experiment
 ├╴examples/demo.rs:73:18
 ├╴Unable to set up experiments
 │
 ├─▶ invalid experiment description
 │   ├╴examples/demo.rs:43:10
 │   ╰╴Experiment 2 could not be parsed
 │
 ╰┬▶ invalid digit found in string
  │  ├╴examples/demo.rs:25:18
  │  ╰╴"3o" could not be parsed as experiment
  │
  ╰▶ invalid digit found in string
     ├╴examples/demo.rs:25:18
     ╰╴"4a" could not be parsed as experiment

 Experiment error: Could not run experiment
 ├╴examples/demo.rs:67:17
 ├╴Experiment 3 has no valid description
 ╰╴Unable to set up experiments

> ```

Reviewer 4: Capitalization is wrong here

Author: Thanks, fixed in 0e3f7f9ff183c9e0e45df867c8293e81fb1f7419

Again, a huge thanks to Bilal Mahmoud ([@indietyp](https://github.com/indietyp)) for his work.

## Improved documentation, `Termination`, `anyhow` and `eyre`

We’ve made a few other improvements too:

- The `error-stack` documentation has been greatly improved and we’re always open to more feedback ([create an Issue](https://github.com/hashintel/hash/tree/main/packages/libs/error-stack) or [join us in Discord](https://discord.com/invite/BPMrGAhjPh)).
- `Termination` has been implemented for `Report`s: it will return the `ExitCode` provided by any `Context` if present, or `ExitCode::FAILURE` will be returned.

> - Compatibility for `anyhow` and `eyre` has been added to convert their types into `Report`.

Reviewer: Should we mention this in the second paragraph?

Author: I think that this is a good idea unless @Alfred-Mountfield has an argument against it (it cannot be done from an external trait).

Reviewer 2: I was convinced this was in a prior release (the initial impl was merged back in July https://github.com/hashintel/hash/pull/763). Turns out it wasn't in the initial release, we 100% should mention this in the intro I think.

Author: Added in 231d8025cb155270d39489b25080952af0d383bd

Reviewer 2: Looks good. I still think we should consider 
> maybe emphasise that this should help people test out error-stack, as well as helping with gradual migration.

Author: Hmm, I think this would be placed in the section the comment was added at?

Author: 103d7f7048b3dde851c40fa9fbb37bc706334964

Reviewer 2: [Suggested a change](https://github.com/hashintel/hash/pull/1146/files#r985667589), thanks

>   - We’ve provided a trait to convert `Result<_, anyhow::Error>` or `Result<_, eyre::Report>` to `Result<_, Report<anyhow::Error>>` or `Result<_, Report<eyre::Report>>`.

Author: Putting these in bullets has the same issues 😞

>   - We’ve provided a trait to convert `Result<_, anyhow::Error>` or `Result<_, eyre::Report>` to `Result<_, Report<anyhow::Error>>` or `Result<_, Report<eyre::Report>>`.

Reviewer 2: I think it's due to the additional nesting. If we found a way to get these back to the initial indentation it'd be fine.

Not sure it's worth us trying to solve this.

>   - We’ve provided a trait to convert `Result<_, anyhow::Error>` or `Result<_, eyre::Report>` to `Result<_, Report<anyhow::Error>>` or `Result<_, Report<eyre::Report>>`.

Reviewer 3: Captured as [a ticket](https://app.asana.com/0/1202953520344821/1203084649340035/f) to look at this week, but marketing as OOS for this PR itself.

- Remove the unused features `hooks`, `futures`, and `futures-core`
- `Report::backtrace` (and `Report::span_trace` for consistency) were deprecated because you cannot get `Backtrace` from `Error` anymore.
  - non-nightly channels should use `Report::downcast_ref::<Backtrace>` / `Report::downcast_ref::<SpanTrace>`.
  - nightly channels should use`Report::requested_ref::<Backtrace>` / `Report::requested_ref::<SpanTrace>`.
- The format of the `Debug` output has changed significantly. If you had a dependency on the format of the `Debug` output (which we don’t recommend), it might break.
- We deprecated `IntoReport::report`. Please use `IntoReport::into_report` instead.
- We deprecated the following as part of the changes we made to how hooks work:
  - instead of `Report::set_debug_hook` please use `Report::install_debug_hook`.
  - Please don’t use `Report::set_display_hook`.
- Because multiple error sources are now possible, we deprecated:
  - `Frame::source`; please use `Frame::sources` instead.
  - `Frame::source_mut`; please use `Frame::sources_mut` instead.

## What’s next?

> We’ve got a lot of ideas! Among them are: adding support for `serde` to provide a hook-based interface to serialize `Report`s; adding support for `defmt`; adding support for WASM environments; and continuing to iterate on the ergonomics of how we support related errors and multiple sources. Please let us know what you think.

Reviewer: Optional bulleting

Where should people let us know? Can we link somewhere?
```suggestion
We’ve got a lot of ideas! Among them are:
- adding support for `serde` to provide a hook-based interface to serialize `Report`s.
- adding support for `defmt`; adding support for WASM environments.
- and continuing to iterate on the ergonomics of how we support related errors and multiple sources. 

Please let us know what you think!
```

Author: I think bullets are more readable.

I think the "Getting Involved" section below is close enough to this section, or what do you think?

Author: 5be57d54f25ad98d028464178618020e3c0c3c33

---
authorPhoto: blog/authors/tim-diekmann.jpg
postPhoto: blog/10_error-stack-update-0-2/Cover.jpg
title: "Announcing error-stack v0.2"

> subtitle: "Redesigned and customizable outputs, related errors and multiple sources, and much more"

Reviewer 3: ```suggestion
subtitle: "Introducing customizable outputs, related errors, multiple sources, and much more to error handling in Rust"
```

We should avoid having 'and' three times in the same sentence. I have also expanded the description slightly (open to feedback).

author: "Tim Diekmann"
jobTitle: "Platform Engineer, HASH"
date: "2022-10-03"
---

> Back in June [we announced the first version](https://hash.dev/blog/announcing-error-stack) of `error-stack`, our context-aware error library. We developed this to help address a lot of challenges we encountered when implementing error handling in our [simulation engine]. In the time since then we’ve started using it in our entity-graph query layer that we’re building as a datastore for [HASH]. Across that development we’ve learned a lot of lessons, and in joint efforts with open-source contributors we’re ready to release a new version of the library.

Reviewer 3: ```suggestion
Back in June [we announced the initial public release](https://hash.dev/blog/announcing-error-stack) of `error-stack`, our context-aware Rust error library. We developed this to help address a lot of challenges we encountered when implementing error handling in our [simulation engine]. In the time since then we’ve started using it in our entity-graph query layer that we’re building as a datastore for [HASH]. Across that development we’ve learned a lot of lessons, and in joint efforts with open-source contributors we’re ready to release a new version of the library.
```

> Since our last release we’ve been fixing user-reported bugs, reviewing open-source contributor’s pull requests, and adding new features. [Version 0.2](https://github.com/hashintel/hash/tree/error-stack%400.2.0/packages/libs/error-stack) is available now and includes important changes from `std`, a completely redesigned output with new customization options, support for related errors and multiple error sources, compatibility with `anyhow` and `eyre`, and a number of other improvements.

Reviewer 3: @TimDiekmann Do we want to link to this GitHub page, or the. `crates.io` listing?

Author: I changed that link to point to crates.io and added a link to GitHub as well - similar as we did in the initial blog post: f171bbbc94a6fec6844d4317cbe541562d881e76

[simulation engine]: https://hash.ai/platform/engine
[hash]: https://hash.ai

> ## Changes from `std` in v0.2

Reviewer 3: ```suggestion
## What's new?

### Changes from `std` in v0.2
```

- The `Provider` API has landed and replaces our vendored-in implementation.
- `Backtrace` was stabilized with Rust 1.65(🎉): we now support `Backtrace`s on a non-nightly channel starting with 1.65-beta.
- If the root context already contains a backtrace, no new `Backtrace` is captured. Otherwise, backtraces are captured automatically when `RUST_BACKTRACE` or `RUST_LIB_BACKTRACE` env-vars are set.
- All std-`Error` types now implement `Provider`. The blanket implementation on `Context` will re-use the `Error::provide` values.
- `Error` is now available in `core::error`, for nightly channels.

> ## A redesigned output with a hook interface for customization

Reviewer 3: ```suggestion
### A redesigned output with a hook interface for customization
```

```text
experiment error: could not run experiment
 ├╴examples/demo.rs:67:17
 ├╴backtrace with 19 frames (1)
 ├╴experiment 3 has no valid description
 ╰╴unable to set up experiments

 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

 Backtrace No. 1
    0: std::backtrace_rs::backtrace::libunwind::trace
              at /rustc/8b705839cd656d202e920efa8769cbe43a5ee269/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
    1: std::backtrace_rs::backtrace::trace_unsynchronized
              at /rustc/8b705839cd656d202e920efa8769cbe43a5ee269/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
    2: std::backtrace::Backtrace::create
              at /rustc/8b705839cd656d202e920efa8769cbe43a5ee269/library/std/src/backtrace.rs:333:13
    3: core::ops::function::FnOnce::call_once
              at /rustc/8b705839cd656d202e920efa8769cbe43a5ee269/library/core/src/ops/function.rs:251:5
    ...
    18: _main
```

For a full guide to using hooks, please refer to [the `fmt` module documentation](https://docs.rs/error-stack/0.2.0/error_stack/fmt/index.html).

> ## Related errors and multiple error sources

Reviewer 3: ```suggestion
### Related errors and multiple error sources
```

`error-stack` v0.2 introduces support for related errors and multiple error sources, which can be invaluable when multiple errors occur at the same time, e.g. in loops or in multi-threaded applications. `Report`s are no longer a linear stream of events. You can now create a tree of different `Report`s, enabling the capture of multiple simultaneous errors without losing information.

This is also shown in the output:

```text
Error: experiment error: could not run experiment
 ├╴examples/demo.rs:73:18
 ├╴Unable to set up experiments
 │
 ├─▶ invalid experiment description
 │   ├╴examples/demo.rs:43:10
 │   ╰╴experiment 2 could not be parsed
 │
 ╰┬▶ invalid digit found in string
  │  ├╴examples/demo.rs:25:18
  │  ╰╴"3o" could not be parsed as experiment
  │
  ╰▶ invalid digit found in string
     ├╴examples/demo.rs:25:18
     ╰╴"4a" could not be parsed as experiment

 experiment error: could not run experiment
 ├╴examples/demo.rs:67:17
 ├╴experiment 3 has no valid description
 ╰╴unable to set up experiments
```

Again, a huge thanks to Bilal Mahmoud ([@indietyp](https://github.com/indietyp)) for his work.

> ## Improved documentation, `Termination`, `anyhow` and `eyre`

Reviewer 3: ```suggestion
### Improved documentation, `Termination`, `anyhow` and `eyre`
```

We’ve made a few other improvements too:

- The `error-stack` documentation has been greatly improved and we’re always open to more feedback ([create an Issue](https://github.com/hashintel/hash/tree/main/packages/libs/error-stack) or [join us in Discord](https://discord.com/invite/BPMrGAhjPh)).
- `Termination` has been implemented for `Report`s: it will return the `ExitCode` provided by any `Context` if present, or `ExitCode::FAILURE` will be returned.
- Compatibility for `anyhow` and `eyre` has been added to convert their types into `Report`.

  - We’ve provided a trait to convert `Result<_, anyhow::Error>` or `Result<_, eyre::Report>` to `Result<_, Report<anyhow::Error>>` or `Result<_, Report<eyre::Report>>`.
  - The additional sources of `anyhow::Error` and `eyre::Report` are also attached as frames to maintain their history.

>   This will help testing out `error-stack` in existing code bases and will also simplify migrating to `error-stack` is desired.

Reviewer 2: ```suggestion
  This should make it much easier to try out `error-stack` in existing codebases which utilize these libraries, as well as improving the migration process if desired.
```

> ## Breaking changes

Reviewer 3: ```suggestion
### Breaking changes
```

- Remove the unused features `hooks`, `futures`, and `futures-core`
- `Report::backtrace` (and `Report::span_trace` for consistency) were deprecated because you cannot get `Backtrace` from `Error` anymore.
  - non-nightly channels should use `Report::downcast_ref::<Backtrace>` / `Report::downcast_ref::<SpanTrace>`.
  - nightly channels should use`Report::requested_ref::<Backtrace>` / `Report::requested_ref::<SpanTrace>`.
- The format of the `Debug` output has changed significantly. If you had a dependency on the format of the `Debug` output (which we don’t recommend), it might break.
- We deprecated `IntoReport::report`. Please use `IntoReport::into_report` instead.
- We deprecated the following as part of the changes we made to how hooks work:
  - instead of `Report::set_debug_hook` please use `Report::install_debug_hook`.
  - Please don’t use `Report::set_display_hook`.
- Because multiple error sources are now possible, we deprecated:
  - `Frame::source`; please use `Frame::sources` instead.
  - `Frame::source_mut`; please use `Frame::sources_mut` instead.

## What’s next?

> We’ve got a lot of ideas! Among them are:

Reviewer 3: ```suggestion
We’ve got lots of new features planned, including:
```

- The `error-stack` documentation has been greatly improved and we’re always open to more feedback ([create an Issue](https://github.com/hashintel/hash/tree/main/packages/libs/error-stack) or [join us in Discord](https://discord.com/invite/BPMrGAhjPh)).
- `Termination` has been implemented for `Report`s: it will return the `ExitCode` provided by any `Context` if present, or `ExitCode::FAILURE` will be returned.
- Compatibility for `anyhow` and `eyre` has been added to convert their types into `Report`.

  - We’ve provided a trait to convert `Result<_, anyhow::Error>` or `Result<_, eyre::Report>` to `Result<_, Report<anyhow::Error>>` or `Result<_, Report<eyre::Report>>`.
  - The additional sources of `anyhow::Error` and `eyre::Report` are also attached as frames to maintain their history.

>   This will help testing out `error-stack` in existing code bases and will also simplify migrating to `error-stack` is desired.

Reviewer: ```suggestion
  This will help testing out `error-stack` in existing code bases and will also simplify migrating to `error-stack`.
```

Reviewer 2: See this comment for alternative: https://github.com/hashintel/hash/pull/1146#discussion_r985667589

Author: Applied in 38d1f9007ef83eb2cb28cd82b16b9e8e531e87f8

Reviewer 2: I would have kept it in the new phrasing, but don't feel strongly.

Author: I think this will simplify migrating in any case, even if users don't want to use it 🙂
